### ISSUE TITLE
apiserver manually populates endpoints for the "kubernetes" Service, but does not consider its own readiness

### ISSUE THREAD

**Issue by fusida** [kubernetes/kubernetes#119535]
### What happened?

We have an apiserver with kms. if kms (or any other) probe fail, the kubernetes services endpoints still exist in default namespaces.

curl /readyz
[-]kms-provider-0 failed: reason withheld

curl /api/v1/namespaces/default/endpoints/kubernetes
"subsets": [
        {
          "addresses": [
            {
              "ip": "xxx.xxx.xxx.xxx"
            }
          ],
          ……
        }
      ]

### What did you expect to happen?

If all of the apiserver pod fail to readiness probe, the client-go in-cluster way can not connect apiserver pod using kubernetes endpoints.

### How can we reproduce it (as minimally and precisely as possible)?

I think the kubernetes endpoints subset need to be update once the apiserver pod is probe fail. It's better to do in server side than client side.

### Kubernetes version

1.22,1.24,1.26
<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by alexzielenski** [comment]
/sig node

may also be interested in this

**Comment by alexzielenski** [comment]
/cc @aojea 
I believe you have some context on the k8s service?

**Comment by mmiranda96** [comment]
Doesn't look related to node.

/remove-sig node
/sig network

**Comment by aojea** [comment]
Interesting ... the endpoints reconciler assumes that the apiserver, once it is ready, it is always ready ... it seems that we have to add a feedback loop to be able to provide this functionality

One thing to consider is that this will add more complexity into the endpoints logic and a dependency on the apiserver availability on an external component that  may cause new problems, thinking out loud, an external component like the kms provider may have a cyclic dependency on the apiserver readyz ...

cc @wojtek-t @deads2k @liggitt @enj @sttts

**Comment by jyotimahapatra** [comment]
some nuances to consider: the endpoints should fail open when all apiservers fail kms check at the same time. 

Also, I'd be curious to learn if the apiserver setup in your case has a livez probe that can depend on `livez` . kms check failure in such cases can lead to apiserver restart leading to re-checking readyz. In my setup, kms check failing for a longer time indicates a bug(failure in cleaning the uniz domain socket) and warrants a restart to auto mitigate.
Also, the usecase if legit, but should capture fail open scenario and should be backward compatible. The usecase in point is that certain users create large amount of secrets and as an operator i cannot ask them not to. The solution should make sure that such users don't see a cascading error because removing the k8s endpoint will cause all traffic to skew to remaining servers(usually 1 or 2 in my case). If we did this change, we should not backport it to release branches.

**Comment by fusida** [comment]
> some nuances to consider: the endpoints should fail open when all apiservers fail kms check at the same time.
> 
> I'd be curious to learn if the apiserver setup in your case has a livez probe that can depend on `livez` . kms check is part of livez check and a failure in such cases can lead to apiserver restart leading to re-checking readyz. In my setup, kms check failing for a longer time indicates a bug(failure in cleaning the unix domain socket) and warrants a restart to auto mitigate. Also, the usecase if legit, but should capture fail open scenario and should be backward compatible. The usecase in point is that certain users create large amount of secrets and as an operator i cannot ask them not to. The solution should make sure that such users don't see a cascading error because removing the k8s endpoint will cause all traffic to skew to remaining servers(usually 1 or 2 in my case). If we did this change, we should not backport it to release branches.

@jyotimahapatra if restart apiserver, will cause relist from etcd to init apiserver cache, what's more in larger cluster the ETCD are under a lot of pressure. so we won't to restart frequently. we put readyz 1min timeout, livez 2min timeout and etcd exclude.  we hope when readyz fail caused by kms, the default service can remove the apiserver endpoints instead of relying on apiserver restart (we would like to set livez kms exclude) .

**Comment by fusida** [comment]
> 
Thanks for your answer, if kms provider is not deployed within the cluster it should not depend on the default kubernetes service, nor on whether apiserver is ready（apiserver kms sidecar is client, kms provider is server）, and the default kubernetes endpoints reconciler should comply with the readyz。

**Comment by khenidak** [comment]
> Interesting ... the endpoints reconciler assumes that the apiserver, once it is ready, it is always ready ... it seems that we have to add a feedback loop to be able to provide this functionality

for that reason @aojea  we should revisit this. Consider a networking partition where one api-server lost connection to etcd. but still available to nodes. traffic shouldn't be routed there to that unhealthy api-server.

**Comment by jyotimahapatra** [comment]
>  Consider a networking partition where one api-server lost connection to etcd. but still available to nodes

I feel this is handled cleanly since the endpoints are deleted when etcd ttl expires. The error mode may rather happen when apiserver to etcd connectivity is intact, but the health/ready probes are failing.

**Comment by aojea** [comment]
> for that reason @aojea we should revisit this. Consider a networking partition where one api-server lost connection to etcd. but still available to nodes. traffic shouldn't be routed there to that unhealthy api-server.

> The error mode may rather happen when apiserver to etcd connectivity is intact, but the health/ready probes are failing.

Right, I agree this is wrong, there is not feedback loop from the readiness to the endpoint reconciler

**Comment by fusida** [comment]
> > Consider a networking partition where one api-server lost connection to etcd. but still available to nodes
> 
> I feel this is handled cleanly since the endpoints are deleted when etcd [ttl](https://github.com/kubernetes/kubernetes/blob/a0873d37c58757ac6c2a339af1e4f8d601cdef20/pkg/controlplane/reconcilers/lease.go#L97) expires. The error mode may rather happen when apiserver to etcd connectivity is intact, but the health/ready probes are failing.
 
Reliance on etcd timeout notifications: Not every probe type failure will cause apiserver not to update the lease.
Restart apiserver depending on the livez probe: livez and readyz can be configured with different exclude, not all readyz will restart.

I agree with this :endpoints reconciler should take into account ready detection results, and preferably remove traffic in a timely manner.
@jyotimahapatra @aojea @khenidak

**Comment by aojea** [comment]
/assign

I may have some time to work on this, the key is on how to plumb down the readiness information to the reconciler

**Comment by aojea** [comment]
Ok, I think this will fix this problem https://github.com/kubernetes/kubernetes/pull/121105, it will be nice to have feedback of this patch in real environments

**Comment by aojea** [comment]
> Ok, I think this will fix this problem #121105, it will be nice to have feedback of this patch in real environments

that PR got closed and is not clear a change that impact the API server healthz , that is so critical, will be added with a more larger discussions

/unassign

**Comment by thockin** [comment]
As discussed here  - this is a very dangerous area.  It's very easy to end up with a bad dependency killing the whole intra-cluster control-plane.  At the same time, there are cases where apiservers may be legitimately partitioned.

Realistically, it's not something that a single apiserver alone can know.  It can always reach itself - it needs outside opinions.  At the storage layer we have quorum, so a partitioned API should fail at that level, right?

We do not have a fail-open model WRT Endpoints or EndpointSlices - both can represent "unready" endpoints but we don't DO much with that.  There is a "fallback on terminating-ready endpoints if no running-ready endpoints exist" but no "fallback on running-unready endpoints if no running-ready endpoints exist".

I don't think we know how to do this safely yet, and AFAICT this is not a bad enough that anyone is looking at it in the near future.
